crates/sui/src/upgrade_compatibility.rs

```diff
@@ -353,6 +359,56 @@ impl CompatibilityMode for CliCompatibilityMode {
     }
 }
 
+/// Generates an enum Category along with individual enum for each individual category
+/// and impls into diagnostic info for each category.
+macro_rules! upgrade_codes {
+    ($($cat:ident: [
+        $($code:ident: { msg: $code_msg:literal }),* $(,)?
+    ]),* $(,)?) => {
+        #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash, PartialOrd, Ord)]
+        #[repr(u8)]
+        pub enum Category {
+            #[allow(dead_code)]
+            ZeroPlaceholder,
+            $($cat,)*
+        }
+
+        $(
+            #[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
+            #[repr(u8)]
+            pub enum $cat {
+                #[allow(dead_code)]
+                ZeroPlaceholder,
+                $($code,)*
+            }
+
+            // impl into diagnostic info
+            impl Into<DiagnosticInfo> for $cat {
+                fn into(self) -> DiagnosticInfo {
+                    match self {
+                        Self::ZeroPlaceholder =>
+                            panic!("do not use placeholder error code"),
+                        $(Self::$code => custom(
+                            "COMPATIBILITY",
```

Comment:
```suggestion
                            "Compatibility ",
```

Reply:
And we might want to bind this to a constant somewhere. `const COMPATIBILITY_PREFIX: &str` or something similar.